cirq/experiments/t2_decay_experiment.py

```diff
@@ -132,52 +169,117 @@ def t2_decay(
             study.Points('inv_x', [0.0, -0.5]),
             study.Points('inv_y', [-0.5, 0.0]),
         )
-        sweep = study.Product(delay_sweep, tomography_sweep)
-    elif experiment_type == ExperimentType.HAHN_ECHO:
-        # Hahn / Spin Echo T2 experiment
-        # Use sqrt(Y) to flip to the equator.
-        # Evolve the state for half the given amount of delay time
-        # Flip the state using an X gate
-        # Evolve the state for half the given amount of delay time
-        # Then measure the state in both X and Y bases.
-
-        circuit = circuits.Circuit(
-            ops.Y(qubit)**0.5,
-            ops.WaitGate(value.Duration(nanos=0.5 * delay_var))(qubit),
-            ops.X(qubit),
-            ops.WaitGate(value.Duration(nanos=0.5 * delay_var))(qubit),
-            ops.X(qubit)**inv_x_var,
-            ops.Y(qubit)**inv_y_var,
-            ops.measure(qubit, key='output'),
-        )
+    else:
+        if experiment_type == ExperimentType.HAHN_ECHO:
+            # Hahn / Spin Echo T2 experiment
+            # Use sqrt(Y) to flip to the equator.
+            # Evolve the state for the given amount of delay time
+            # Flip the state using an X gate
+            # Evolve the state for the given amount of delay time
+            # Then measure the state in both X and Y bases.
+            num_pulses = [0]
+            # This is equivalent to a CPMG experiment with zero pulses
+            # and will follow the same code path.
+
+        # Carr-Purcell-Meiboom-Gill sequence.
+        # Performs the following sequence
+        # π/2 - wait(t) - π - wait(2t) - ... - π - wait(t)
+        # There will be N π pulses (X gates)
+        # where N sweeps over the values of num_pulses
+        #
+        if not num_pulses:
+            raise ValueError('At least one value must be given '
+                             'for num_pulses in a CPMG experiment')
+        circuit = _cpmg_circuit(qubit, delay_var, max_pulses)
+        circuit.append(ops.X(qubit)**inv_x_var)
```

Comment:
This part:

```python
circuit.append(ops.X(qubit)**inv_x_var)
circuit.append(ops.Y(qubit)**inv_y_var)
circuit.append(ops.measure(qubit, key='output'))
tomography_sweep = study.Zip(
    study.Points('inv_x', [0.0, 0.5]),
    study.Points('inv_y', [-0.5, 0.0]),
)
```

is state tomography and AFAIU is independent of which type of experiment is performed. It is also duplicated above for Ramsey experiment (modulo a minus sign which does not seem to matter).

Perhaps I'm missing something, but I don't see why we need to implement tomography twice in this function. If indeed we don't then please reduce the indent from here until line 200 and remove lines 164-166 and 168-171.

This will reduce code verbosity and make the code better reflect what is happening here by introducing a clear separation between the core of each T2 experiment and state tomography (which is just a generic method of characterizing quantum states).

---

It would be even better if we did not implement tomography here at all and instead called [`cirq. single_qubit_state_tomography`](https://github.com/quantumlib/Cirq/blob/master/cirq/experiments/qubit_characterizations.py#L357). WDYT?

Reply:
Let's defer. Filed #3316.

Reply:
I'll leave using cirq. single_qubit_state_tomography to another PR.  I moved the tomography outside the conditional.